### DISCUSSION TITLE
get drho_dT from Water97FluidProperties

### DISCUSSION THREAD

**Discussion by SunJianhang678** [idaholab/moose#29016] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [x] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear MOOSE experts,
I'm using the Water97FluidProperties and trying to get drho_dT as a property of my material so that it could be consumed by other kernels. But I don't know how to achieve it. I want to get drho_dT in a way like how i get rho, by declaring _rho in .h file and getting it in .C file, but it seems that Water97Fluidproperties does not have a member named 'drho_dT_from_p_T' corresponding to 'rho_from_p_T'. Part of my current code is as below.
Could you please offer some guidance? It will be great help if there are some examples using drho_dT this way. Thank you so much!

.h file
private:  
  MaterialProperty<Real> & _rho;
  MaterialProperty<Real> & _drho_dT;
  MaterialProperty<Real> & _drho_dp;
  const Water97FluidProperties & _fluid;

.C file
Water97::Water97(const InputParameters & parameters)
  : Material(parameters),    
    _rho(declareProperty<Real>("rho")),
    _drho_dT(declareProperty<Real>("drho_dT")),
    _drho_dp(declareProperty<Real>("drho_dp")),
    _fluid(getUserObject<Water97FluidProperties>("fp"))
{
}

void
Water97::computeQpProperties()
{
  _rho[_qp] = _fluid.rho_from_p_T(_p[_qp], _T[_qp]);
  _drho_dT[_qp] = _fluid.drho_dT_from_p_T(_p[_qp], _T[_qp]);
  _drho_dp[_qp] = _fluid.drho_dp_from_p_T(_p[_qp], _T[_qp]);  
}

**Comment by SunJianhang678** [comment]
I solved this problem by replacing the last two lines showed above with the following codes: 
  double delta = 1e-1; 
  double rho_dT = _fluid.rho_from_p_T(_p[_qp], _T[_qp] + delta);
  double rho_dp = _fluid.rho_from_p_T(_p[_qp] + delta, _T[_qp]);

  _drho_dT[_qp] = (rho_dT - _rho[_qp]) / delta;
  _drho_dp[_qp] = (rho_dp - _rho[_qp]) / delta;

but I'm wondering if there were more convenient ways provided by moose, instead of calculating the values by myself?

**Reply by GiudGiud** [reply, marked as answer]
This is fine but the convenient way is to do this:
Finite differencing will be a slightly more imprecise than using the routines that compute those derivatives
```
Real rho, drho_dT, drho_dp;
_fluid.rho_from_p_T(_p[_qp], _T[_qp], rho, drho_dT, drho_dp);
```

**Reply by SunJianhang678** [reply]
Got it! When I read the source code of Water97FP, I didn't know why more params than needed appear in this function and what they are for, now I understand. Thank you so much!

**Reply by GiudGiud** [reply]
p, T are inputs
rho, drho_dT, drho_dp are references to a number. So while they are passed as arguments, they are actually outputs of this routine

so this routine officially returns "void" but in practice it computes rho, drho_dT, drho_dp in the (reference to) numbers you pass as parameters
